### ISSUE TITLE
[BUG] NPE when scheduling recuring

### ISSUE THREAD

**Issue by mendess** [jobrunr/jobrunr#182]
**Are you running the latest version of JobRunr?**
Yes

**Describe the bug**

Null pointer exception is thrown by `scheduleRecurring`

**Environment**
I'm using JobRunr version: 3.2.1
I'm running on JRE / JDK (e.g. OpenJDK 1.8.0_292): 11
I'm using the following StorageProvider (e.g. Oracle / MongoDB / ...): postgres

**How to Reproduce**
I first wrote the code in kotlin but changed to java to check if the problem persisted. It does.

```java
import org.jobrunr.JobRunrException;
import org.jobrunr.scheduling.JobScheduler;
import org.springframework.beans.factory.ListableBeanFactory;

import java.time.ZoneId;

public class JobRunrHack {
    public static void schedule(
            JobScheduler jobScheduler,
            String jobId,
            String cron,
            ZoneId zoneId,
            String beanName,
            String methodName
    ) {
        try {
            jobScheduler.<ListableBeanFactory>scheduleRecurrently(jobId, cron, zoneId, l -> {
                final var bean = l.getBean(beanName);
                bean.getClass().getMethod(methodName).invoke(bean);
            });
        } catch (JobRunrException e) {
            e.printStackTrace();
            throw e;
        }
    }
}
```

**Expected behavior**
No output and job is scheduled

**Additional context**
```
org.jobrunr.JobRunrException: JobRunr encountered a problematic exception. Please create a bug report (if possible, provide the code to reproduce this and the stacktrace)
	at org.jobrunr.JobRunrException.shouldNotHappenException(JobRunrException.java:43)
	at org.jobrunr.jobs.details.JobDetailsGeneratorUtils.createObjectViaMethod(JobDetailsGeneratorUtils.java:70)
	at org.jobrunr.jobs.details.instructions.JobDetailsInstruction.getObject(JobDetailsInstruction.java:79)
	at org.jobrunr.jobs.details.instructions.JobDetailsInstruction.invokeInstruction(JobDetailsInstruction.java:44)
	at org.jobrunr.jobs.details.instructions.AbstractJVMInstruction.invokeInstructionAndPushOnStack(AbstractJVMInstruction.java:18)
	at org.jobrunr.jobs.details.JobDetailsBuilder.invokeInstructions(JobDetailsBuilder.java:92)
	at org.jobrunr.jobs.details.JobDetailsBuilder.getJobDetails(JobDetailsBuilder.java:70)
	at org.jobrunr.jobs.details.AbstractJobDetailsFinder.getJobDetails(AbstractJobDetailsFinder.java:84)
	at org.jobrunr.jobs.details.JavaJobDetailsFinder.getJobDetails(JavaJobDetailsFinder.java:10)
	at org.jobrunr.jobs.details.JobDetailsAsmGenerator.toJobDetails(JobDetailsAsmGenerator.java:27)
	at org.jobrunr.scheduling.JobScheduler.scheduleRecurrently(JobScheduler.java:561)
	at com.emitu.cloud.core.tasks.JobRunrHack.schedule(JobRunrHack.java:19)
```

From some debugging I found that `JobDetailsInstruction::getObject` passes null to `createObjectiViaMethod` this is `null` is obtained from the stack because it was put there by https://github.com/jobrunr/jobrunr/blob/4feb7f5655d1b36e9330e69ad867f7cb95d0c024/core/src/main/java/org/jobrunr/jobs/details/JavaJobDetailsBuilder.java#L31-L33

**Comment by mendess** [comment]
During this "investigation" I found that a new `@Recurring` annotation is being developed, this is exactly what I was trying to implement myself, which lead me to this problem

**Comment by rdehuyss** [comment]
@mendess the next release will be 4.0 and I hope to release it somewhere in September (I'm also taking some holidays).
It will also include better support for Quarkus and Micronaut.

Regarding your bug: there is a page on best practices. Try to minimize the amount of work within the job lambda.

**Comment by mendess** [comment]
I did read the page on best practices and I tried to follow it. The lambda captures only 2 strings and it calls just a couple of methods. Could the use of reflection be an issue? I changed the code to be:
```java
import lombok.extern.slf4j.Slf4j;
import org.jobrunr.JobRunrException;
import org.jobrunr.scheduling.JobScheduler;
import org.springframework.beans.factory.ListableBeanFactory;

import java.lang.reflect.InvocationTargetException;
import java.time.ZoneId;

@Slf4j
public class JobRunrHack {
    public static void schedule(
            JobScheduler jobScheduler,
            String jobId,
            String cron,
            ZoneId zoneId,
            String beanName,
            String methodName
    ) {
        try {
            jobScheduler.<ListableBeanFactory>scheduleRecurrently(
                    jobId,
                    cron,
                    zoneId,
                    l -> _run(l, beanName, methodName)
            );
        } catch (JobRunrException e) {
            e.printStackTrace();
            throw e;
        }
    }

    public static void _run(ListableBeanFactory listableBeanFactory, String beanName, String methodName) {
        final var bean = listableBeanFactory.getBean(beanName);
        try {
            bean.getClass().getMethod(methodName).invoke(bean);
        } catch (IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
            log.error("Failed to call " + methodName + " on " + beanName, e);
        }
    }
}
```
and I still get a NPE, this time inside `Autoboxer::autobox`, but it seems to be the same problem, that null that is inserted in the stack is not handled correctly.

**Comment by zhk0603** [comment]
I also met

**Comment by rdehuyss** [comment]
I invested this and I see what happens - this is a use case I did not think about and I also don't think I will support. I will update the documentation though.

The reason it does not work is that the generic type (in this case `ListableBeanFactory`) is to be used when executing the job (as in the examples on the website). In this case, JobRunr asks for the `ListableBeanFactory` from the Spring Application Context (which is in fact itself) and then executes a method on it. Here, it is used during the analysis of the job (while enqueueing), it is not available and thus the `NullPointerException`.

What you are trying to achieve is however possible (and with less code):

Just inject the `ListableBeaFactory` or `ApplicationContext` instead of passing and use that to resolve the bean.

```
package org.jobrunr.examples.api;

import org.jobrunr.JobRunrException;
import org.jobrunr.examples.services.SampleJobService;
import org.jobrunr.jobs.JobId;
import org.jobrunr.scheduling.JobScheduler;
import org.springframework.beans.factory.ListableBeanFactory;
import org.springframework.context.ApplicationContext;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.time.Duration;
import java.time.ZoneId;
import java.util.UUID;

import static java.time.Instant.now;

@RestController
public class JobController {

    private final JobScheduler jobScheduler;
    private final ApplicationContext applicationContext;

    public JobController(JobScheduler jobScheduler, ApplicationContext applicationContext) {
        this.jobScheduler = jobScheduler;
        this.applicationContext = applicationContext;
    }

    @GetMapping("/schedule-issue")
    public void scheduleIssue() {
        schedule(jobScheduler,"some-recurring-job", "*/15 * * * *", ZoneId.systemDefault(), "sampleJobService", "recurringJob");
    }

    public void schedule(
            JobScheduler jobScheduler,
            String jobId,
            String cron,
            ZoneId zoneId,
            String beanName,
            String methodName
    ) {
        try {
            jobScheduler.scheduleRecurrently(
                    jobId,
                    cron,
                    zoneId,
                    x -> _run(beanName, methodName)
            );
        } catch (JobRunrException e) {
            e.printStackTrace();
            throw e;
        }
    }

    public void _run(String beanName, String methodName) {
        final Object bean = applicationContext.getBean(beanName);
        try {
            bean.getClass().getMethod(methodName).invoke(bean);
        } catch (Exception e) {
            System.out.println("Failed to call " + methodName + " on " + beanName);
            e.printStackTrace();
        }
    }
}
```

Does this make sense?

**Comment by mendess** [comment]
> `x -> _run(beanName, methodName)`
what is the type of `x` in this lambda?

**Comment by RahulSDeshpande** [comment]
@mendess & @rdehuyss

I am strangely facing similar issue since yesterday.

I don't know what is wrong going, I didn't change any line of code in this file since long.

Here is my code:
```kt
@PostConstruct
fun startJobRunr() {
    jobScheduler.scheduleRecurrently(Cron.minutely()) {
        runJob1()
        runJob2()
    }
}

@Job
fun runJob2() {
    ..
}

@Job
fun runJob2() {
    ..
}
```

And here is my crash log stack-trace:
```kt
Caused by: org.jobrunr.JobRunrException: JobRunr encountered a problematic exception. Please create a bug report (if possible, provide the code to reproduce this and the stacktrace)
	at org.jobrunr.JobRunrException.shouldNotHappenException(JobRunrException.java:43) ~[jobrunr-3.2.1.jar:na]
	at org.jobrunr.jobs.details.JobDetailsGeneratorUtils.createObjectViaMethod(JobDetailsGeneratorUtils.java:70) ~[jobrunr-3.2.1.jar:na]
	at org.jobrunr.jobs.details.instructions.JobDetailsInstruction.getObject(JobDetailsInstruction.java:79) ~[jobrunr-3.2.1.jar:na]
	at org.jobrunr.jobs.details.instructions.JobDetailsInstruction.invokeInstruction(JobDetailsInstruction.java:44) ~[jobrunr-3.2.1.jar:na]
	at org.jobrunr.jobs.details.instructions.AbstractJVMInstruction.invokeInstructionAndPushOnStack(AbstractJVMInstruction.java:18) ~[jobrunr-3.2.1.jar:na]
	at org.jobrunr.jobs.details.JobDetailsBuilder.invokeInstructions(JobDetailsBuilder.java:92) ~[jobrunr-3.2.1.jar:na]
	at org.jobrunr.jobs.details.JobDetailsBuilder.getJobDetails(JobDetailsBuilder.java:70) ~[jobrunr-3.2.1.jar:na]
	at org.jobrunr.jobs.details.AbstractJobDetailsFinder.getJobDetails(AbstractJobDetailsFinder.java:84) ~[jobrunr-3.2.1.jar:na]
	at org.jobrunr.jobs.details.KotlinJobDetailsFinder.getJobDetails(KotlinJobDetailsFinder.java:16) ~[jobrunr-3.2.1.jar:na]
	at org.jobrunr.jobs.details.JobDetailsAsmGenerator.toJobDetails(JobDetailsAsmGenerator.java:16) ~[jobrunr-3.2.1.jar:na]
	at org.jobrunr.scheduling.JobScheduler.scheduleRecurrently(JobScheduler.java:542) ~[jobrunr-3.2.1.jar:na]
	at org.jobrunr.scheduling.JobScheduler.scheduleRecurrently(JobScheduler.java:506) ~[jobrunr-3.2.1.jar:na]
	at org.jobrunr.scheduling.JobScheduler.scheduleRecurrently(JobScheduler.java:472) ~[jobrunr-3.2.1.jar:na]
	at ******* HIDDEN AS ITS A PRIVATE PROJECT FILE PATH **.startJobRunr(ScheduledTasks.kt:97) ~[classes/:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_282]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_282]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_282]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_282]
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleElement.invoke(InitDestroyAnnotationBeanPostProcessor.java:389) ~[spring-beans-5.3.5.jar:5.3.5]
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleMetadata.invokeInitMethods(InitDestroyAnnotationBeanPostProcessor.java:333) ~[spring-beans-5.3.5.jar:5.3.5]
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor.postProcessBeforeInitialization(InitDestroyAnnotationBeanPostProcessor.java:157) ~[spring-beans-5.3.5.jar:5.3.5]
`

Is there any way I could solve this!?

Cheers 👯

**Comment by rdehuyss** [comment]
I had to debug it also to find out :smile: In this case, it was the `JobController` itself.

> `() -> _run(beanName, methodName)`

 would have been a better example.

**Comment by rdehuyss** [comment]
Solved in 4.0.0-RC2 (somewhere tomorrow in Maven Central) and in updated docs once deployed
